### PR TITLE
Leaderboard rework

### PR DESCRIPTION
**Pull request by rbischof**
- Fixes bug that only one player with same amount of trophies gets displayed in leaderboard

- Instead of re-fetching all friends multiple times, they get fetched once and cached. The filter is then applied to the cached list.

- All private helper classes were moved into their own files (CC complained) and everything put into a leaderboard folder.

- Added a class TestUsers which contains the IDs of all users created for testing purposes on Firebase. You can now call TestUsers.isTestUser(userId) to check if the userId belongs to one of the testUsers.

- Solved a bunch of CC issues

### CODE REVIEW

File: app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+        entry.setBackgroundColor(res.getColor(
+                isCurrentUser ? R.color.colorDrawYellow : R.color.colorLightGrey));
+        entry.setPadding(30, 10, 30, 10);
+
+        return entry;
+    }
+
+    private LinearLayout addViews(LinearLayout layout, View[] views) {
+        for (int i = 0; i < views.length; ++i) {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java]: Here you can use a for-each

File: app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        view.setText(text);
+        view.setTextSize(20);
+        view.setMaxLines(1);
+        view.setTextColor(color);
+        Typeface typeMuro = Typeface.createFromAsset(context.getAssets(), "fonts/Muro.otf");
+        view.setTypeface(typeMuro);
+        view.setLayoutParams(layoutParams);
+    }
+
+    void setRank(int rank) {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java]: Is it used only for testing? We could add the VisibleForTesting annotation if it is the case

File: app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package ch.epfl.sweng.SDP.utils;
+
+public class TestUsers {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java]: Add final and a javadoce here ;)

File: app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    /**
+     * Checks if the given id belongs to a user created only for testing purposes.
+     *
+     * @param userId    to check
+     * @return          true if id belongs to a test user, else false
+     * @throws          IllegalArgumentException if userId is null or empty
+     */
+    public static boolean isTestUser(String userId) {
+        if (userId == null || userId.isEmpty()) {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java]: You can replace this if with checkPrecondition

File: app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            throw new IllegalArgumentException("UserId must not be null or empty");
+        }
+
+        for (int i = 0; i < allTestUsers.length; ++i) {
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java]: You can use a for-each here

File: app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java (hunk 1 above)
Comment (rbischof) [reply app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java]: I did already try, but forEach was only introduced in java8 i think?

Comment (Gondolav) [reply app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java]: I think it was introduced in Java 5! It doesn't work?

Comment (rbischof) [reply app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java]: Ah I see which forEach you mean. You're right.

File: app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java (hunk 2 above)
Comment (rbischof) [reply app/src/main/java/ch/epfl/sweng/SDP/home/leaderboard/Player.java]: This is not my function, but as far as I understood this is used for setting the rank before the username (e.g. 1. PICASSO, 2. GYRO, etc). So it must be package private.

File: app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package ch.epfl.sweng.SDP.utils;
+
+import static ch.epfl.sweng.SDP.utils.Preconditions.checkPrecondition;
+
+/**
+ * Contains the IDs of all users on Firebase that were created for testing purposes.
+ * Implements a method for checking ig a given user is a real or a test user.
```
Comment (Gondolav) [inline app/src/main/java/ch/epfl/sweng/SDP/utils/TestUsers.java]: if ;)

**Pull request merged**